File: Application/src/main/java/com/shareconnect/ServiceApiClient.java
package com.shareconnect;

import android.util.Log;
import okhttp3.*;
import org.json.JSONObject;
import java.io.IOException;

public class ServiceApiClient {
    private static final String TAG = "ServiceApiClient";
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    
    private OkHttpClient client;
    
    public ServiceApiClient() {
        client = new OkHttpClient();
    }
    
    public interface ServiceApiCallback {
        void onSuccess();
        void onError(String error);
    }
    
    /**
     * Send URL to the appropriate service based on profile type
     */
    public void sendUrlToService(ServerProfile profile, String url, ServiceApiCallback callback) {
        switch (profile.getServiceType()) {
            case ServerProfile.TYPE_METUBE:
                sendUrlToMeTube(profile, url, callback);
                break;
            case ServerProfile.TYPE_YTDL:
                sendUrlToYtdl(profile, url, callback);
                break;
            case ServerProfile.TYPE_TORRENT:
                sendUrlToTorrentClient(profile, url, callback);
                break;
            case ServerProfile.TYPE_JDOWNLOADER:
                sendUrlToJDownloader(profile, url, callback);
                break;
            default:
                callback.onError("Unsupported service type: " + profile.getServiceType());
                break;
        }
    }
    
    /**
     * Send URL to MeTube service
     */
    private void sendUrlToMeTube(ServerProfile profile, String youtubeUrl, ServiceApiCallback callback) {
        try {
            // Construct the MeTube API endpoint
            String url = profile.getUrl() + ":" + profile.getPort() + "/add";
            
            // Create the JSON payload
            JSONObject json = new JSONObject();
            json.put("url", youtubeUrl);
            json.put("quality", "best"); // Default to best quality
            
            RequestBody body = RequestBody.create(json.toString(), JSON);
            Request request = new Request.Builder()
                    .url(url)
                    .post(body)
                    .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Failed to send URL to MeTube", e);
                    callback.onError(e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            callback.onSuccess();
                        } else {
                            String errorBody = response.body() != null ? response.body().string() : "Unknown error";
                            Log.e(TAG, "MeTube API error: " + errorBody);
                            callback.onError("API Error: " + response.code() + " - " + errorBody);
                        }
                    } finally {
                        if (response.body() != null) {
                            response.body().close();
                        }
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error preparing MeTube API request", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send URL to YT-DLP service
     */
    private void sendUrlToYtdl(ServerProfile profile, String url, ServiceApiCallback callback) {
        try {
            // Construct the YT-DLP API endpoint (similar to MeTube)
            String apiUrl = profile.getUrl() + ":" + profile.getPort() + "/add";
            
            // Create the JSON payload
            JSONObject json = new JSONObject();
            json.put("url", url);
            json.put("quality", "best"); // Default to best quality
            
            RequestBody body = RequestBody.create(json.toString(), JSON);
            Request request = new Request.Builder()
                    .url(apiUrl)
                    .post(body)
                    .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Failed to send URL to YT-DLP", e);
                    callback.onError(e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            callback.onSuccess();
                        } else {
                            String errorBody = response.body() != null ? response.body().string() : "Unknown error";
                            Log.e(TAG, "YT-DLP API error: " + errorBody);
                            callback.onError("API Error: " + response.code() + " - " + errorBody);
                        }
                    } finally {
                        if (response.body() != null) {
                            response.body().close();
                        }
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error preparing YT-DLP API request", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send URL to torrent client
     */
    private void sendUrlToTorrentClient(ServerProfile profile, String url, ServiceApiCallback callback) {
        try {
            String baseUrl = profile.getUrl() + ":" + profile.getPort();
            
            // Check if it's a magnet link or regular URL
            if (url.startsWith("magnet:")) {
                // For magnet links, send directly to torrent client
                sendMagnetToTorrentClient(profile, url, callback);
            } else {
                // For regular URLs, we might need to handle differently depending on the client
                sendRegularUrlToTorrentClient(profile, url, callback);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error preparing torrent client request", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send regular URL to torrent client
     */
    private void sendRegularUrlToTorrentClient(ServerProfile profile, String url, ServiceApiCallback callback) {
        try {
            String baseUrl = profile.getUrl() + ":" + profile.getPort();
            String apiUrl = "";
            
            // Different torrent clients have different APIs
            switch (profile.getTorrentClientType()) {
                case ServerProfile.TORRENT_CLIENT_QBITTORRENT:
                    // qBittorrent Web API
                    apiUrl = baseUrl + "/api/v2/torrents/add";
                    sendToQBittorrent(apiUrl, url, callback);
                    break;
                    
                case ServerProfile.TORRENT_CLIENT_TRANSMISSION:
                    // Transmission RPC API
                    apiUrl = baseUrl + "/transmission/rpc";
                    sendToTransmission(apiUrl, url, callback);
                    break;
                    
                case ServerProfile.TORRENT_CLIENTUTORRENT:
                    // uTorrent Web API
                    apiUrl = baseUrl + "/gui/";
                    sendToUTorrent(apiUrl, url, callback);
                    break;
                    
                default:
                    callback.onError("Unsupported torrent client: " + profile.getTorrentClientType());
                    break;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error sending URL to torrent client", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send magnet link to torrent client
     */
    private void sendMagnetToTorrentClient(ServerProfile profile, String magnetUrl, ServiceApiCallback callback) {
        try {
            String baseUrl = profile.getUrl() + ":" + profile.getPort();
            String apiUrl = "";
            
            // Different torrent clients have different APIs
            switch (profile.getTorrentClientType()) {
                case ServerProfile.TORRENT_CLIENT_QBITTORRENT:
                    // qBittorrent Web API
                    apiUrl = baseUrl + "/api/v2/torrents/add";
                    sendToQBittorrent(apiUrl, magnetUrl, callback);
                    break;
                    
                case ServerProfile.TORRENT_CLIENT_TRANSMISSION:
                    // Transmission RPC API
                    apiUrl = baseUrl + "/transmission/rpc";
                    sendToTransmission(apiUrl, magnetUrl, callback);
                    break;
                    
                case ServerProfile.TORRENT_CLIENTUTORRENT:
                    // uTorrent Web API
                    apiUrl = baseUrl + "/gui/";
                    sendToUTorrent(apiUrl, magnetUrl, callback);
                    break;
                    
                default:
                    callback.onError("Unsupported torrent client: " + profile.getTorrentClientType());
                    break;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error sending magnet to torrent client", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send to qBittorrent
     */
    private void sendToQBittorrent(String apiUrl, String url, ServiceApiCallback callback) {
        try {
            // qBittorrent expects form data
            RequestBody formBody = new MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart("urls", url)
                    .build();
            
            Request request = new Request.Builder()
                    .url(apiUrl)
                    .post(formBody)
                    .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Failed to send to qBittorrent", e);
                    callback.onError(e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            callback.onSuccess();
                        } else {
                            String errorBody = response.body() != null ? response.body().string() : "Unknown error";
                            Log.e(TAG, "qBittorrent API error: " + errorBody);
                            callback.onError("API Error: " + response.code() + " - " + errorBody);
                        }
                    } finally {
                        if (response.body() != null) {
                            response.body().close();
                        }
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error preparing qBittorrent request", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send to Transmission
     */
    private void sendToTransmission(String apiUrl, String url, ServiceApiCallback callback) {
        try {
            // Transmission expects JSON-RPC
            JSONObject json = new JSONObject();
            json.put("method", "torrent-add");
            JSONObject params = new JSONObject();
            
            // Transmission can handle both URLs and magnet links
            if (url.startsWith("magnet:")) {
                params.put("filename", url);
            } else {
                params.put("filename", url);
            }
            
            json.put("arguments", params);
            
            RequestBody body = RequestBody.create(json.toString(), JSON);
            Request request = new Request.Builder()
                    .url(apiUrl)
                    .post(body)
                    .addHeader("X-Transmission-Session-Id", "dummy") // May need to handle session ID properly
                    .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Failed to send to Transmission", e);
                    callback.onError(e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            callback.onSuccess();
                        } else {
                            String errorBody = response.body() != null ? response.body().string() : "Unknown error";
                            Log.e(TAG, "Transmission API error: " + errorBody);
                            callback.onError("API Error: " + response.code() + " - " + errorBody);
                        }
                    } finally {
                        if (response.body() != null) {
                            response.body().close();
                        }
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error preparing Transmission request", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send to uTorrent
     */
    private void sendToUTorrent(String apiUrl, String url, ServiceApiCallback callback) {
        try {
            // uTorrent Web API - this is a simplified version
            String fullUrl = apiUrl + "?action=add-url&s=" + url;
            
            Request request = new Request.Builder()
                    .url(fullUrl)
                    .get()
                    .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Failed to send to uTorrent", e);
                    callback.onError(e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            callback.onSuccess();
                        } else {
                            String errorBody = response.body() != null ? response.body().string() : "Unknown error";
                            Log.e(TAG, "uTorrent API error: " + errorBody);
                            callback.onError("API Error: " + response.code() + " - " + errorBody);
                        }
                    } finally {
                        if (response.body() != null) {
                            response.body().close();
                        }
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error preparing uTorrent request", e);
            callback.onError(e.getMessage());
        }
    }
    
    /**
     * Send URL to jDownloader
     */
    private void sendUrlToJDownloader(ServerProfile profile, String url, ServiceApiCallback callback) {
        try {
            // jDownloader My.JDownloader API or direct API
            String baseUrl = profile.getUrl() + ":" + profile.getPort();
            String apiUrl = baseUrl + "/flashget"; // Common jDownloader endpoint for direct downloads
            
            // jDownloader typically accepts GET requests with URL parameters
            String fullUrl = apiUrl + "?url=" + url;
            
            Request request = new Request.Builder()
                    .url(fullUrl)
                    .get()
                    .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    Log.e(TAG, "Failed to send to jDownloader", e);
                    callback.onError(e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        if (response.isSuccessful()) {
                            callback.onSuccess();
                        } else {
                            String errorBody = response.body() != null ? response.body().string() : "Unknown error";
                            Log.e(TAG, "jDownloader API error: " + errorBody);
                            callback.onError("API Error: " + response.code() + " - " + errorBody);
                        }
                    } finally {
                        if (response.body() != null) {
                            response.body().close();
                        }
                    }
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error preparing jDownloader request", e);
            callback.onError(e.getMessage());
        }
    }
}

File: Application/src/main/kotlin/com/shareconnect/ServiceApiClient.kt
package com.shareconnect

import com.redelf.commons.logging.Console
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException

class ServiceApiClient {
    private val client: OkHttpClient

    interface ServiceApiCallback {
        fun onSuccess()
        fun onError(error: String?)
    }

    init {
        // Initialize client with cookie support for proper authentication handling
        client = OkHttpClient.Builder()
            .cookieJar(okhttp3.CookieJar.NO_COOKIES) // We'll handle authentication manually
            .build()
    }

    /**
     * Send URL to the appropriate service based on profile type
     */
    fun sendUrlToService(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        when (profile.serviceType) {
            ServerProfile.TYPE_METUBE -> sendUrlToMeTube(profile, url, callback)
            ServerProfile.TYPE_YTDL -> sendUrlToYtdl(profile, url, callback)
            ServerProfile.TYPE_TORRENT -> sendUrlToTorrentClient(profile, url, callback)
            ServerProfile.TYPE_JDOWNLOADER -> sendUrlToJDownloader(profile, url, callback)
            else -> callback.onError("Unsupported service type: " + profile.serviceType)
        }
    }

    /**
     * Send URL to MeTube service
     */
    private fun sendUrlToMeTube(profile: ServerProfile, youtubeUrl: String, callback: ServiceApiCallback) {
        try {
            // Construct the MeTube API endpoint
            val url = profile.url + ":" + profile.port + "/add"

            // Create the JSON payload
            val json = JSONObject()
            json.put("url", youtubeUrl)
            json.put("quality", "best") // Default to best quality

            val body: RequestBody = json.toString().toRequestBody(JSON)
            val requestBuilder = Request.Builder().url(url).post(body)
            
            // Add authentication if provided
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send URL to MeTube")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("MeTube API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing MeTube API request")
            callback.onError(e.message)
        }
    }

    /**
     * Send URL to YT-DLP service
     */
    private fun sendUrlToYtdl(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        try {
            // Construct the YT-DLP API endpoint (similar to MeTube)
            val apiUrl = profile.url + ":" + profile.port + "/add"

            // Create the JSON payload
            val json = JSONObject()
            json.put("url", url)
            json.put("quality", "best") // Default to best quality

            val body: RequestBody = json.toString().toRequestBody(JSON)
            val requestBuilder = Request.Builder().url(apiUrl).post(body)
            
            // Add authentication if provided
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send URL to YT-DLP")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("YT-DLP API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing YT-DLP API request")
            callback.onError(e.message)
        }
    }

    /**
     * Send URL to torrent client
     */
    private fun sendUrlToTorrentClient(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        try {
            val baseUrl = profile.url + ":" + profile.port

            // Check if it's a magnet link or regular URL
            if (url.startsWith("magnet:")) {
                // For magnet links, send directly to torrent client
                sendMagnetToTorrentClient(profile, url, callback)
            } else {
                // For regular URLs, we might need to handle differently depending on the client
                sendRegularUrlToTorrentClient(profile, url, callback)
            }
        } catch (e: Exception) {
            Console.error(e, "Error preparing torrent client request")
            callback.onError(e.message)
        }
    }

    /**
     * Send regular URL to torrent client
     */
    private fun sendRegularUrlToTorrentClient(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        try {
            val baseUrl = profile.url + ":" + profile.port
            val apiUrl: String

            // Different torrent clients have different APIs
            when (profile.torrentClientType) {
                ServerProfile.TORRENT_CLIENT_QBITTORRENT -> {
                    // qBittorrent Web API
                    apiUrl = baseUrl + "/api/v2/torrents/add"
                    sendToQBittorrent(profile, apiUrl, url, callback)
                }
                ServerProfile.TORRENT_CLIENT_TRANSMISSION -> {
                    // Transmission RPC API
                    apiUrl = baseUrl + "/transmission/rpc"
                    sendToTransmission(profile, apiUrl, url, callback)
                }
                ServerProfile.TORRENT_CLIENTUTORRENT -> {
                    // uTorrent Web API
                    apiUrl = baseUrl + "/gui/"
                    sendToUTorrent(profile, apiUrl, url, callback)
                }
                else -> callback.onError("Unsupported torrent client: " + profile.torrentClientType)
            }
        } catch (e: Exception) {
            Console.error(e, "Error sending URL to torrent client")
            callback.onError(e.message)
        }
    }

    /**
     * Send magnet link to torrent client
     */
    private fun sendMagnetToTorrentClient(profile: ServerProfile, magnetUrl: String, callback: ServiceApiCallback) {
        try {
            val baseUrl = profile.url + ":" + profile.port
            val apiUrl: String

            // Different torrent clients have different APIs
            when (profile.torrentClientType) {
                ServerProfile.TORRENT_CLIENT_QBITTORRENT -> {
                    // qBittorrent Web API
                    apiUrl = baseUrl + "/api/v2/torrents/add"
                    sendToQBittorrent(profile, apiUrl, magnetUrl, callback)
                }
                ServerProfile.TORRENT_CLIENT_TRANSMISSION -> {
                    // Transmission RPC API
                    apiUrl = baseUrl + "/transmission/rpc"
                    sendToTransmission(profile, apiUrl, magnetUrl, callback)
                }
                ServerProfile.TORRENT_CLIENTUTORRENT -> {
                    // uTorrent Web API
                    apiUrl = baseUrl + "/gui/"
                    sendToUTorrent(profile, apiUrl, magnetUrl, callback)
                }
                else -> callback.onError("Unsupported torrent client: " + profile.torrentClientType)
            }
        } catch (e: Exception) {
            Console.error(e, "Error sending magnet to torrent client")
            callback.onError(e.message)
        }
    }

    /**
     * Send to qBittorrent
     */
    private fun sendToQBittorrent(profile: ServerProfile, apiUrl: String, url: String, callback: ServiceApiCallback) {
        try {
            // qBittorrent expects form data
            val formBody: RequestBody = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("urls", url)
                .build()

            val requestBuilder = Request.Builder().url(apiUrl).post(formBody)
            
            // Add authentication if provided
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send to qBittorrent")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("qBittorrent API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing qBittorrent request")
            callback.onError(e.message)
        }
    }

    /**
     * Send to Transmission
     */
    private fun sendToTransmission(profile: ServerProfile, apiUrl: String, url: String, callback: ServiceApiCallback) {
        try {
            // Transmission expects JSON-RPC
            val json = JSONObject()
            json.put("method", "torrent-add")
            val params = JSONObject()

            // Transmission can handle both URLs and magnet links
            params.put("filename", url)

            json.put("arguments", params)

            val body: RequestBody = json.toString().toRequestBody(JSON)
            val requestBuilder = Request.Builder().url(apiUrl).post(body)
            
            // Add authentication if provided
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send to Transmission")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("Transmission API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing Transmission request")
            callback.onError(e.message)
        }
    }

    /**
     * Send to uTorrent
     */
    private fun sendToUTorrent(profile: ServerProfile, apiUrl: String, url: String, callback: ServiceApiCallback) {
        try {
            // uTorrent Web API - improved implementation with proper token handling
            // First, we need to get a token from /token.html
            val tokenUrl = "${profile.url}:${profile.port}/gui/token.html"
            
            val tokenRequestBuilder = Request.Builder().url(tokenUrl).get()
            // Add authentication for token request
            addAuthentication(tokenRequestBuilder, profile)
            
            val tokenRequest = tokenRequestBuilder.build()

            client.newCall(tokenRequest).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to get uTorrent token")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            val responseBody = response.body?.string()
                            // Extract token from response (simplified - in practice would parse HTML)
                            val token = extractTokenFromResponse(responseBody ?: "")
                            
                            // Now send the URL with the token
                            sendUrlToUTorrentWithToken(profile, url, token, callback)
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("uTorrent token error: $errorBody")
                            callback.onError("Token Error: " + response.code + " - $errorBody")
                        }
                    } catch (e: Exception) {
                        Console.error(e, "Error processing uTorrent token response")
                        callback.onError(e.message)
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing uTorrent request")
            callback.onError(e.message)
        }
    }

    /**
     * Send URL to uTorrent with token
     */
    private fun sendUrlToUTorrentWithToken(profile: ServerProfile, url: String, token: String, callback: ServiceApiCallback) {
        try {
            val baseUrl = "${profile.url}:${profile.port}"
            // uTorrent Web API - add URL with token
            val fullUrl = "$baseUrl/gui/?action=add-url&s=$url&token=$token"

            val requestBuilder = Request.Builder().url(fullUrl).get()
            // Add authentication for the actual request
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send to uTorrent")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("uTorrent API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error sending URL to uTorrent")
            callback.onError(e.message)
        }
    }

    /**
     * Extract token from uTorrent token response
     * Simplified implementation - in practice would parse HTML properly
     */
    private fun extractTokenFromResponse(response: String): String {
        // Simplified token extraction - in a real implementation, 
        // this would properly parse the HTML to extract the token
        val tokenRegex = "<div id='token' style='display:none;'>([^<]+)</div>".toRegex()
        val matchResult = tokenRegex.find(response)
        return matchResult?.groupValues?.getOrNull(1) ?: ""
    }

    /**
     * Send URL to jDownloader
     */
    private fun sendUrlToJDownloader(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        try {
            // Try My.JDownloader API first (modern approach)
            sendToMyJDownloader(profile, url, object : ServiceApiCallback {
                override fun onSuccess() {
                    callback.onSuccess()
                }
                
                override fun onError(error: String?) {
                    // If My.JDownloader fails, fall back to legacy direct API
                    sendToLegacyJDownloader(profile, url, callback)
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing jDownloader request")
            callback.onError(e.message)
        }
    }

    /**
     * Send URL to My.JDownloader API (modern approach)
     */
    private fun sendToMyJDownloader(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        try {
            val baseUrl = "${profile.url}:${profile.port}"
            // My.JDownloader API endpoint for adding links
            val apiUrl = "$baseUrl/flash/add"

            // Prepare JSON payload for My.JDownloader
            val json = JSONObject()
            json.put("urls", arrayOf(url))
            json.put("packageName", "") // Optional package name
            json.put("destinationFolder", "") // Optional destination folder
            
            val body: RequestBody = json.toString().toRequestBody(JSON)
            
            val requestBuilder = Request.Builder().url(apiUrl).post(body)
            
            // Add authentication if provided
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send to My.JDownloader")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("My.JDownloader API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing My.JDownloader request")
            callback.onError(e.message)
        }
    }

    /**
     * Send URL to legacy jDownloader direct API (fallback approach)
     */
    private fun sendToLegacyJDownloader(profile: ServerProfile, url: String, callback: ServiceApiCallback) {
        try {
            val baseUrl = "${profile.url}:${profile.port}"
            // Legacy jDownloader direct API endpoint
            val apiUrl = "$baseUrl/flashget" // Common jDownloader endpoint for direct downloads

            // jDownloader typically accepts GET requests with URL parameters
            val fullUrl = "$apiUrl?url=$url"

            val requestBuilder = Request.Builder().url(fullUrl).get()
            
            // Add authentication if provided
            addAuthentication(requestBuilder, profile)
            
            val request = requestBuilder.build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Console.error(e, "Failed to send to legacy jDownloader")
                    callback.onError(e.message)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    try {
                        if (response.isSuccessful) {
                            callback.onSuccess()
                        } else {
                            val errorBody = if (response.body != null) response.body!!.string() else "Unknown error"
                            Console.error("Legacy jDownloader API error: $errorBody")
                            callback.onError("API Error: " + response.code + " - $errorBody")
                        }
                    } finally {
                        response.body?.close()
                    }
                }
            })
        } catch (e: Exception) {
            Console.error(e, "Error preparing legacy jDownloader request")
            callback.onError(e.message)
        }
    }

    /**
     * Helper method to add authentication to a request builder
     */
    private fun addAuthentication(requestBuilder: Request.Builder, profile: ServerProfile) {
        if (!profile.username.isNullOrEmpty() && !profile.password.isNullOrEmpty()) {
            // Basic authentication
            val credentials = okhttp3.Credentials.basic(profile.username!!, profile.password!!)
            requestBuilder.addHeader("Authorization", credentials)
        }
    }

    companion object {
        private const val TAG = "ServiceApiClient"
        private val JSON: MediaType = "application/json; charset=utf-8".toMediaType()
    }
}